src/OptimizedPathMappingRepository.php

```diff
@@ -0,0 +1,457 @@
+    private function addResource($path, FilesystemResource $resource)
+    {
+        // Don't modify resources attached to other repositories
+        if ($resource->isAttached()) {
+            $resource = clone $resource;
+        }
+
+        // Read children before attaching the resource to this repository
+        $children = $resource->listChildren();
+
+        $resource->attachTo($this, $path);
+
+        // Add the resource before adding its children, so that the array stays sorted
+        $this->store->set($path, $resource->getFilesystemPath());
+
+        $basePath = '/' === $path ? $path : $path.'/';
+
+        foreach ($children as $name => $child) {
+            $this->addResource($basePath.$name, $child);
+        }
+    }
+
+
+    /**
+     * @param string $path
+     */
+    private function removePath($path)
+    {
+        if (!$this->store->exists($path)) {
+            return;
+        }
+
+        // Remove children first
+        $children = $this->getRecursivePathChildIterator($path);
+
+        foreach ($children as $child) {
+            $this->store->remove($child);
+        }
+
+        // Remove the resource
+        $this->store->remove($path);
+    }
+
+    /**
+     * Returns an iterator for the children paths of a resource.
+     *
+     * @param Resource $resource The resource.
+     *
+     * @return RegexFilterIterator|string[] The iterator of paths.
+     */
+    private function getChildIterator(Resource $resource)
+    {
+        $staticPrefix = rtrim($resource->getPath(), '/').'/';
+        $regExp = '~^'.preg_quote($staticPrefix, '~').'[^/]+$~';
+
+        return new RegexFilterIterator(
+            $regExp,
+            $staticPrefix,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Returns a recursive iterator for the children paths under a given path.
+     *
+     * @param string $path The path.
+     *
+     * @return RegexFilterIterator|string[] The iterator of paths.
+     */
+    private function getRecursivePathChildIterator($path)
+    {
+        $staticPrefix = rtrim($path, '/').'/';
+        $regExp = '~^'.preg_quote($staticPrefix, '~').'.+$~';
+
+        return new RegexFilterIterator(
+            $regExp,
+            $staticPrefix,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Returns an iterator for a glob.
+     *
+     * @param string $glob The glob.
+     *
+     * @return GlobFilterIterator|string[] The iterator of paths.
+     */
+    private function getGlobIterator($glob)
+    {
+        return new GlobFilterIterator(
+            $glob,
+            new ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Recursively creates a directory for a path.
+     *
+     * @param string $path A directory path.
+     * @return DirectoryResource The created resource
+     */
+    private function ensureDirectoryExists($path)
+    {
+        if (!$this->store->exists($path)) {
```

Comment:
Could you invert the condition and add an early return instead? (like in createRoot())